util/chunk/checksum.go

```diff
@@ -0,0 +1,134 @@
+package chunk
+
+import (
+	"encoding/binary"
+	"errors"
+	"hash/crc32"
+	"io"
+	"sync"
+)
+
+const (
+	checksumBlockSize   = 1024
+	checksumSize        = 4
+	checksumPayloadSize = checksumBlockSize - checksumSize
+)
+
+var checksumReaderBufPool = sync.Pool{
+	New: func() interface{} { return make([]byte, checksumBlockSize) },
+}
+
+// checksumWriter implements an io.WriteCloser, it calculates and stores a CRC-32 checksum for the payload before
+// writing to the underlying object.
+type checksumWriter struct {
+	w           io.WriteCloser
+	buf         []byte
+	payload     []byte
+	payloadUsed int
+}
```

Comment:
Add a field `err error`. We can learn from bufio.Writer how to record and return an error.